File: installer/lib/phases/scan.ts
import { existsSync } from 'node:fs';
import type { Artifact, Manifest } from '@torba/core';
import { filterManifest } from '@torba/core';
import { interpolate } from '@torba/core';
import type { OsOptions } from '@torba/rules';

export interface ScanTask {
  artifact: Artifact;
  finalPath: string;
  idx: number;
}

export interface ScanResult {
  tasks: ScanTask[];
  skipped: number;
}

export function scan(
  manifest: Manifest,
  vars: Record<string, string>,
  platform: OsOptions,
  /**
   * `artifact.path` templates to fetch unconditionally, even if the file
   * already exists — pointer artifacts whose upstream has moved on.
   */
  force: ReadonlySet<string> = new Set(),
): ScanResult {
  const applicable = filterManifest(manifest, platform);
  const tasks: ScanTask[] = [];
  let skipped = 0;

  for (let i = 0; i < applicable.artifacts.length; i++) {
    const u = applicable.artifacts[i]!;
    const finalPath = interpolate(u.path, vars);
    if (!force.has(u.path) && existsSync(finalPath)) {
      skipped++;
      continue;
    }
    tasks.push({ artifact: u, finalPath, idx: i });
  }

  return { tasks, skipped };
}

File: crates/lanka-runtime/src/phases/scan.rs
use std::collections::HashSet;
use std::path::Path;
use lanka_core::{filter_manifest, interpolate, Artifact, Manifest, OsOptions};

use crate::errors::InstallError;

pub struct ScanTask {
    pub artifact: Artifact,
    pub final_path: String,
}

pub struct ScanResult {
    pub tasks: Vec<ScanTask>,
    pub skipped: u32,
}

pub fn scan(
    manifest: &Manifest,
    vars: &indexmap::IndexMap<String, String>,
    platform: &OsOptions,
    feats: &[String],
    force: &HashSet<String>,
) -> Result<ScanResult, InstallError> {
    let applicable = filter_manifest(manifest, platform, feats)?;
    let mut tasks = Vec::new();
    let mut skipped: u32 = 0;

    for u in applicable.artifacts {
        let final_path = interpolate(&u.path, vars);
        if !force.contains(&u.path) && Path::new(&final_path).exists() {
            skipped += 1;
            continue;
        }
        tasks.push(ScanTask {
            artifact: u,
            final_path,
        });
    }

    Ok(ScanResult { tasks, skipped })
}
